tubular/dates.py

```diff
@@ -69,23 +73,25 @@ def check_columns_are_date_or_datetime(
             allowed_types = [*allowed_types, date_type]
 
         for col in self.columns:
-            if pd.api.types.is_datetime64_any_dtype(X[col]):
+            is_datetime = pd.api.types.is_datetime64_any_dtype(X[col])
+            is_date = pd.api.types.infer_dtype(X[col]) == date_type
+            if is_datetime:
                 type_dict[col] = datetime_type
 
-            elif (not datetime_only) and (
-                pd.api.types.infer_dtype(X[col]) == date_type
-            ):
+            elif (not datetime_only) and (is_date):
                 type_dict[col] = date_type
 
             else:
-                msg = f"{self.classname()}: {col} type should be in {allowed_types} but got {X[col].dtype}"
+                col_dtype = date_type if is_date else X[col].dtype
+
+                msg = f"{self.classname()}: {col} type should be in {allowed_types} but got {col_dtype}"
                 raise TypeError(msg)
 
         present_types = set(type_dict.values())
 
         only_datetime_present = present_types == {datetime_type}
```

Comment:
Q: I think this bit of logic might be cleaned up a bit by comparing present_types to set(allowed_types)?

Reply:
good shout, updated here https://github.com/lvgig/tubular/pull/305/commits/ae94c5520a09174cdd3ff8406e2c31782ae0b5fa